### PR TITLE
LogQL: add __line__ function to | line_format template

### LINKED ISSUE #4878: Include line in LogQL template

### PR DESCRIPTION
**Pull request by cyriltovena**
This PRs introduce a new `__line__` function for the LogQL `line_format` pipeline.

This function returns the current log line in case the users wants to enrich it.

Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

**Which issue(s) this PR fixes**:
Fixes #4878

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Review by jeschkies — APPROVED** [review]

File: docs/sources/logql/template_functions.md [hunk 1]
```diff
@@ -23,6 +24,23 @@ Example:
 {{ .path | replace " " "_" | trunc 5 | upper }}
 ```
 
+## __line__
+
+This function return the current log line.
+
+> Added in Loki 2.5
```
Comment (KMiller-Grafana) [inline docs/sources/logql/template_functions.md:31]: 2 changes: 
1. grammar correction
2. now that we have versioned docs, the Loki version number in which this function was added is irrelevant. (A task for another PR: remove all the notices that say what version of Loki a function was added to.)

```suggestion
This function returns the current log line.

```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]

File: pkg/logql/log/fmt.go [hunk 2]
```diff
@@ -95,22 +99,29 @@ func init() {
...
+	currentLine []byte
 }
 
 // NewFormatter creates a new log line formatter from a given text template.
 func NewFormatter(tmpl string) (*LineFormatter, error) {
+	lf := &LineFormatter{
+		buf: bytes.NewBuffer(make([]byte, 4096)),
+	}
+	functionMap[functionLineName] = func() string {
+		return unsafeGetString(lf.currentLine)
```
Comment (owen-d) [inline pkg/logql/log/fmt.go:112]: This doesn't seem thread safe 🤔

**Review by owen-d — COMMENTED** [review]
I like the addition, but it doesn't look thread safe. Is there something I'm missing?

File: pkg/logql/log/fmt.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/logql/log/fmt.go:112]: I can add a test for the race but I think it's fine

Comment (cyriltovena) [reply pkg/logql/log/fmt.go:112]: Well let me rephrase that, it's not thread safe but this whole code of LogQL is never run multiple line at a time.

Comment (owen-d) [reply pkg/logql/log/fmt.go:112]: We can have multiple querier workers which can call this concurrently, no?

Comment (cyriltovena) [reply pkg/logql/log/fmt.go:112]: Each worker initialized this once and execute it line by line.

**Review by KMiller-Grafana — APPROVED** [review]
The documentation of this new function looks good to me now.

File: pkg/logql/log/fmt.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/logql/log/fmt.go:112]: I think we're good @owen-d, I can try it in dev if that would make you more confident.

Comment (owen-d) [reply pkg/logql/log/fmt.go:112]: `functionMap` is globally defined variable, so even if we have multiple workers, this looks like a problem. Am I missing something?

Comment (cyriltovena) [reply pkg/logql/log/fmt.go:112]: Ho yeah I was focused on something like else the log line, I need to copy this map fn

Comment (cyriltovena) [reply pkg/logql/log/fmt.go:116]: Good call it's cloned !

**Comment by cyriltovena** [comment]
@owen-d we should be good now, can you confirm and merge ?

**Review by owen-d — APPROVED** [review]
